command/action.go

```diff
@@ -0,0 +1,412 @@
+		return 1
+	}
+
+	var allocStub *api.AllocationListStub
+	if job != "" {
+		// change from bool condition to string
+		// jobID, ns, err := l.JobIDByPrefix(client, args[0], nil)
+		// if err != nil {
+		// 	l.Ui.Error(err.Error())
+		// 	return 1
+		// }
+
+		ns := "default" // TODO: TEMP
+
+		l.Ui.Output(fmt.Sprintf("job passed: %s", job))
+		l.Ui.Output(fmt.Sprintf("Namespace: %s", ns))
+		l.Ui.Output(fmt.Sprintf("action name: %s", args[0]))
+
+		allocStub, err = getRandomJobAlloc(client, job, ns)
+		if err != nil {
+			l.Ui.Error(fmt.Sprintf("Error fetching allocations: %v", err))
+			return 1
+		}
+		l.Ui.Output(fmt.Sprintf("allocStub: %s", allocStub))
+	} else {
+		// I think irrelevant?
+		// allocID := args[0]
+		// allocs, _, err := client.Allocations().PrefixList(sanitizeUUIDPrefix(allocID))
+		// if err != nil {
+		// 	l.Ui.Error(fmt.Sprintf("Error querying allocation: %v", err))
+		// 	return 1
+		// }
+
+		// if len(allocs) == 0 {
+		// 	l.Ui.Error(fmt.Sprintf("No allocation(s) with prefix or id %q found", allocID))
+		// 	return 1
+		// }
+
+		// if len(allocs) > 1 {
+		// 	out := formatAllocListStubs(allocs, false, shortId)
+		// 	l.Ui.Error(fmt.Sprintf("Prefix matched multiple allocations\n\n%s", out))
+		// 	return 1
+		// }
+
+		// allocStub = allocs[0]
+	}
+
+	q := &api.QueryOptions{Namespace: allocStub.Namespace}
+	alloc, _, err := client.Allocations().Info(allocStub.ID, q)
+	if err != nil {
+		l.Ui.Error(fmt.Sprintf("Error querying allocation: %s", err))
+		return 1
+	}
+
+	if task != "" {
+		err = validateTaskExistsInAllocation(task, alloc)
+	} else {
+		task, err = lookupAllocTask(alloc)
+	}
+	if err != nil {
+		l.Ui.Error(err.Error())
+		return 1
+	}
+	l.Ui.Output(fmt.Sprintf("Task set: %s", task))
+
+	if !stdinOpt {
+		l.Stdin = bytes.NewReader(nil)
+	}
+
+	if l.Stdin == nil {
+		l.Stdin = os.Stdin
+	}
+
+	if l.Stdout == nil {
+		l.Stdout = os.Stdout
+	}
+
+	if l.Stderr == nil {
+		l.Stderr = os.Stderr
+	}
+
+	action := args[0]
+	// Convert action to command here.
+	// TODO: Noting that this is also being (redundantly) done in client/alloc_endpoint.go's execImpl.
+	// I could pass the action name all the way down, but that would be maybe polluting the param space of
+	// api/allocations.go, api/allocations_exec.go, etc., since alloc exec normally would never use an Action
+	// except by via this command.
+	// TODO: update: I think I'll try that anyway.
+
+	code, err := l.execImpl(client, alloc, task, job, action, ttyOpt, escapeChar, l.Stdin, l.Stdout, l.Stderr)
+	if err != nil {
+		l.Ui.Error(fmt.Sprintf("failed to exec into task: %v", err))
+		return 1
+	}
+
+	return code
+}
+
+// execImpl invokes the Alloc Exec api call, it also prepares and restores terminal states as necessary.
+func (l *ActionCommand) execImpl(client *api.Client, alloc *api.Allocation, task string, job string, action string, tty bool,
+	escapeChar string, stdin io.Reader, stdout, stderr io.WriteCloser) (int, error) {
```

Comment:
The `execImpl` method in `command/alloc_exec.go` can be made into a free function because it doesn't use anything from its own `(l *AllocExecCommand)`. I would do that and then call it from `Run` above instead of copying all this code (and then trying to get away with dropping features).